### PR TITLE
Improved handling of missing dependencies and support scenarios where users re-create the same env

### LINKED ISSUE #9104: Wrong interpreter is used when installing ipykernel
Issue Type: <b>Bug</b>

I'm trying to run a cell with a new python intepreter

Firslty, it prompts to install ipykernel for python 3.8.5

image

but then it tries to run script against python 2.7 

```
➜  notebooks git:(main) ✗ /usr/bin/python /Users/penlv/.vscode-insiders/extensions/ms-python.python-2022.1.1869134244-dev/pythonFiles/shell_exec.py /usr/bin/pyt
hon -m ensurepip /var/folders/j7/py1_06vs0y3bbcdpsn0pd66h0000gn/T/tmp-26474Q2seHTAsqJTY.log
Executing command in shell >> /usr/bin/python -m ensurepip
/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/OpenSSL/crypto.py:14: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in the next release.
  from cryptography import utils, x509
DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support
Looking in links: /var/folders/j7/py1_06vs0y3bbcdpsn0pd66h0000gn/T/tmpCk4wYC
Requirement already satisfied: setuptools in /System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python (41.0.1)
Collecting pip
Installing collected packages: pip
ERROR: Could not install packages due to an EnvironmentError: [Errno 13] Permission denied: '/Library/Python/2.7/site-packages/pip-19.2.3.dist-info'
Consider using the `--user` option or check the permissions.
```

Extension version: 2022.2.1010551006
VS Code version: Code - Insiders 1.65.0-insider (Universal) (32456dc9e7d3194a94dd3ee0322ccf47e461d09c, 2022-02-23T05:27:54.628Z)
OS version: Darwin x64 21.3.0
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz (16 x 2400)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|4, 4, 5|
|Memory (System)|32.00GB (0.13GB free)|
|Process Argv|--enable-proposed-api ms-vscode.vscode-notebook-history --enable-proposed-api vscode.notebook-editing --enable-proposed-api vscode.vscode-test-resolver --enable-proposed-api donjayamanne.kusto --enable-proposed-api donjayamanne.jupyterpowertools --enable-proposed-api rebornix.better-jupyter-errors --enable-proposed-api rchiodo.jupyter-tools --crash-reporter-id 1a907bbd-fa02-4595-81e0-63e1648afb31|
|Screen Reader|no|
|VM|0%|
</details>

### LINKED ISSUE #9135: IPyKernel not getting installed when re-creating the virtual environment
* Disable zmq support with the following setting
`"jupyter.disableZMQSupport": true`
* Create a virtual env named `xyz`
* Open a notebook, run a cell against this kernel
* Ensure you get a prompt to install ipykernel & things work as expected.
* Now, uninstall ipykernel, reload VS Code and attempt to run the cell & it will not work

### PR DESCRIPTION
**Pull request by DonJayamanne**
Fixes #9135
Fixes #9104

**Here are some of the scenarios fixed (& tests have been added for these & more):**
* User has virtual env `abc`, and ipykernel gets installed into this.
	* User deletes this env and re-creates this, things will no longer work anymore when using non-zmq
* User has global env `/usr/bin/python` pointing to 3.8
	* things work, 
	* Tomorrow they update to 3.9 and `/usr/bin/python` points to 3.9
	* Our kernel picker still displays 3.8 and even when prompted to install ipykernel they see 3.9 in some of the messages
* If user changes kernels a few times, the cells the user ran never get executed.
	* I.e. attempt to run a cell, get prompted to select a kernel, and you get notified that ipykernel doesn't exist.
	* Try this a few times and select different kernels, the cells don't run as expected.
* In some cases when ipykernel is missing you get the prompt to select a kernel, clicking that does nothing (in non-zmq case).

**Added a whole load of tests for a number of scenarios**
* Tests for above scenarios 
* User uninstall ipykernel and attempts to restart the cell, things just hang
* Ensure we don't get pops about missing dependencies when auto starting kernels 
* & more

**Architectural change**
* Previously in Kernel Dependency service class, if user clicks the button to change the kernel, then this class will select a new kernel & controller will re-run the cells.
	* Now cleaner separation, the error handler is responsible for handling the error of ipykernel not installed (previously we had an exception where kernel dependency service would change the kernel & re-run cells).
	* The error handler now notifies others whether cells need to be re-run, much cleaner than using static variables, etc..

### CODE REVIEW

File: src/client/datascience/errors/errorHandler.ts [hunk 1]
```diff
@@ -53,10 +63,20 @@ import { JupyterInterpreterService } from '../jupyter/interpreter/jupyterInterpr
...
+import { JupyterInvalidKernelError } from './jupyterInvalidKernelError';
+import { selectKernel } from '../jupyter/kernels/kernelSelector';
 
+function getFirstCell(cells?: NotebookCell[]) {
+    return cells?.length ? cells[0] : undefined;
+}
 @injectable()
 export class DataScienceErrorHandler implements IDataScienceErrorHandler {
+    private readonly _onShouldRunCells = new EventEmitter<NotebookCell[]>();
+    public readonly onShouldRunCells = this._onShouldRunCells.event;
```
Comment (DonJayamanne) [inline src/client/datascience/errors/errorHandler.ts:79]: Error handler handles the errors now and determines the actions to be taken.
E.g. if ipykernel is not installed, error handler will check if user needs to select a new kernel and once done will then notify other parts of extension whether cells need to be re-run.

File: src/client/datascience/notebook/vscodeNotebookController.ts [hunk 2]
```diff
@@ -292,21 +298,32 @@ export class VSCodeNotebookController implements Disposable {
         // If this NotebookController was selected, fire off the event
         this._onNotebookControllerSelected.fire({ notebook: event.notebook, controller: this });
         this._onNotebookControllerSelectionChanged.fire();
-
-        if (pendingCells?.length) {
-            await this.handleExecution([...pendingCells], event.notebook);
```
Comment (DonJayamanne) [inline src/client/datascience/notebook/vscodeNotebookController.ts:297]: Removed this code, now we handle the `onShouldRunCells` as that's more concrete than using some static variables.

File: src/test/datascience/jupyter/serverSelector.unit.test.ts [hunk 3]
```diff
@@ -111,40 +111,48 @@ suite('DataScience - Jupyter Server URI Selector', () => {
 
     test('Quick pick MRU tests', async () => {
         const { selector, storage } = createDataScienceObject('$(zap) Default', '', true);
-
+        console.log('Step1');
```
Comment (DonJayamanne) [inline src/test/datascience/jupyter/serverSelector.unit.test.ts:114]: For some reason this seems to fail (**sometimes**) on WIndows CI

File: .vscode/settings.json [hunk 4]
```diff
@@ -56,5 +56,9 @@
         // Match what black does.
         "--max-line-length=88"
     ],
-    "typescript.preferences.importModuleSpecifier": "relative"
+    "typescript.preferences.importModuleSpecifier": "relative",
```
Comment (DonJayamanne) [inline .vscode/settings.json:59]: Unit tests will show up in the test explorer now, just copied these settings from the `build/.mocha.unittests.ts.json` file

File: src/client/datascience/notebook/notebookControllerManager.ts [hunk 5]
```diff
@@ -710,7 +711,26 @@ export class NotebookControllerManager implements INotebookControllerManager, IE
...
-                .filter(([id]) => !this.registeredControllers.has(id))
+                .filter(([id]) => {
+                    const controller = this.registeredControllers.get(id);
+                    if (controller) {
+                        // If we already have this controller, its possible the Python version information has changed.
+                        // E.g. we had a cached kernlespec, and since then the user updated their version of python,
+                        // Now we need to update the display name of the kernelspec.
+                        // Assume user created a venv with name `.venv` and points to Python 3.8
+                        // Tomorrow they delete this folder and re-create it with version Python 3.9.
+                        // Similarly they could re-create conda environments or just install a new version of Global Python env.
```
Comment (DonJayamanne) [inline src/client/datascience/notebook/notebookControllerManager.ts:722]: This is one of the issues being addressed

**Review by rchiodo — APPROVED** [review]

File: src/client/datascience/errors/errorHandler.ts (hunk 1 above)
Comment (rchiodo) [reply src/client/datascience/errors/errorHandler.ts:79]: I did this another way. Make the errorHandler return a specific result that says a switch happened. I'll see what effect this has on my change.

**Pull request merged**